src/storage/src/source/metrics.rs

```diff
@@ -196,6 +204,130 @@ impl PostgresSourceSpecificMetrics {
     }
 }
 
+/// Metrics for the `upsert` operator.
+#[derive(Clone, Debug)]
+pub(super) struct UpsertMetrics {
+    pub(super) rehydration_latency: GaugeVec,
+    pub(super) rehydration_total: UIntGaugeVec,
+
+    // These are used by `shared`.
+    pub(super) multi_get_latency: HistogramVec,
+    pub(super) multi_get_size: HistogramVec,
+    pub(super) multi_put_latency: HistogramVec,
+    pub(super) multi_put_size: HistogramVec,
+    // This is a map so that multiple timely workers can interact with the same
+    // `DeleteOnDropHistogram`, which is only dropped once ALL workers drop it.
+    // The map may contain arbitrary, old `Weak`s for deleted sources, which are
+    // only cleaned if those sources are recreated.
+    //
+    // We don't parameterize these by `worker_id` like the `rehydration_*` ones
+    // to save on time-series cardinality.
+    pub(super) shared: Arc<Mutex<BTreeMap<GlobalId, Weak<UpsertSharedMetrics>>>>,
+}
+
+impl UpsertMetrics {
+    fn register_with(registry: &MetricsRegistry) -> Self {
+        let shared = Arc::new(Mutex::new(BTreeMap::new()));
+        Self {
+            rehydration_latency: registry.register(metric!(
+                name: "mz_storage_upsert_state_rehydration_latency",
+                help: "The latency, per-worker, in fractional seconds, \
+                    of rehydrating the upsert state for this source",
+                var_labels: ["source_id", "worker_id"],
+            )),
+            rehydration_total: registry.register(metric!(
+                name: "mz_storage_upsert_state_rehydration_total",
+                help: "The number of values, per-worker, \
+                    rehydrated into the upsert state for this source",
+                var_labels: ["source_id", "worker_id"],
+            )),
+            multi_get_latency: registry.register(metric!(
+                name: "mz_storage_upsert_multi_get_latency",
+                help: "The latencies, in fractional seconds, \
```

Comment:
It sounds like milliseconds would be a better unit for this call

Reply:
Actually, for histograms, fractional seconds are what we use across the codebase!